src/entry.c

```diff
@@ -261,15 +288,49 @@ entry *entryCreate(const_sds field, sds value, long long expiry) {
     bool embed_value = false;
     int embedded_field_sds_type;
     size_t expiry_size, embedded_value_sds_size, embedded_field_sds_size;
-    size_t alloc_size = entryReqSize(field, value, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_sds_size, &expiry_size, &embedded_value_sds_size);
-    size_t buf_size;
+    size_t value_len = value ? sdslen(value) : SIZE_MAX;
+    size_t alloc_size = entryReqSize(sdslen(field), value_len, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_sds_size, &expiry_size, &embedded_value_sds_size);
+    return entryConstruct(alloc_size, field, value, expiry, embed_value, embedded_field_sds_type, expiry_size, embedded_value_sds_size, embedded_field_sds_size);
+}
 
-    /* allocate the buffer */
-    char *buf = zmalloc_usable(alloc_size, &buf_size);
 
-    return entryWrite(buf, buf_size, field, value, expiry, embed_value, embedded_field_sds_type, embedded_field_sds_size, embedded_value_sds_size, expiry_size);
+entry *entryUpdateAsStringRef(entry *entry, const char *buf, size_t len, long long expiry) {
+    long long entry_expiry = entryGetExpiry(entry);
+    // Check for toggling expiration
+    bool expiry_add_remove = (expiry != entry_expiry) && (entry_expiry == EXPIRY_NONE || expiry == EXPIRY_NONE);
+    if (entryHasStringRef(entry) && !expiry_add_remove) {
+        stringRef *value = *entryGetValueRef(entry);
+        value->buf = buf;
+        value->len = len;
+        if (expiry != EXPIRY_NONE) *entryGetExpiryRef(entry) = expiry;
+        return entry;
+    }
+    stringRef *value = zmalloc(sizeof(stringRef));
+    value->buf = buf;
+    value->len = len;
+    if (entryHasValuePtr(entry) && !expiry_add_remove) {
+        sds *value_ref = (sds *)entryGetValueRef(entry);
+        sdsfree(*value_ref);
+        *value_ref = (sds)value;
+        sdsSetAuxBit(entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_STRING_REF, 1);
+        if (expiry != EXPIRY_NONE) *entryGetExpiryRef(entry) = expiry;
+        return entry;
+    }
+    sds field = entryGetField(entry);
+    size_t field_size = sdsReqSize(sdslen(field), SDS_TYPE_8);
+    size_t alloc_size = field_size + sizeof(void *);
+    alloc_size += (expiry == EXPIRY_NONE) ? 0 : sizeof(expiry);
+
+    size_t expiry_size = 0;
+    if (expiry != EXPIRY_NONE) expiry_size = sizeof(expiry);
+    sds new_entry = entryConstruct(alloc_size, field, value, expiry, false, SDS_TYPE_8, expiry_size, sizeof(value), field_size);
```

Comment:
Note the type error.  `sds new_entry`.  This should/could have been caught by the compiler.  This is due to `typedef void entry` in the .h file.  Instead, we should change that to `typedef struct entry entry`.  That will forward declare that entry is some type of struct, allowing use of `entry*`, and preventing unintentional casting.

Now, with `entryConstruct` refactored, this whole chunk simplifies to:
```c
entry *new_entry = entryConstruct(field, NULL, value, expiry);
```
(I'd also rename `value` for clarity to maybe `stringref_value`.)